File: hira/hira/hirapipeline_stack.py
from aws_cdk import (
    Stack,
    SecretValue,
    pipelines,
    aws_iam as iam_,
    aws_codepipeline_actions as cpactions_

)
import aws_cdk as cdk
from constructs import Construct
from .hirastage_stack import HirastageStack

class HirapipelineStack(Stack):

    def __init__(self, scope: Construct, construct_id: str, **kwargs) -> None:
        super().__init__(scope, construct_id, **kwargs)

        role_ = self.create_pipeline_role()

        # Connecting to the github
        source = pipelines.CodePipelineSource.git_hub("hiraaziz2022skipq/sprint3Voyager", "main",
                                                      authentication=SecretValue.secrets_manager('webtken'),
                                                      trigger=cpactions_.GitHubTrigger('POLL'))
        
        synth=pipelines.CodeBuildStep("Synth",
                                    input= source,
                                    commands=["cd hira/", "pip install -r requirements.txt",
                                              "npm install -g aws-cdk", "cdk synth"],
                                    primary_output_directory = "hira/cdk.out",
                                    role = role_)

        # Creating Pipeline
        pipeline = pipelines.CodePipeline(self, "web_Pipeline", synth=synth)

        # Instantiate Beta stage
        beta = HirastageStack(self, "beta")

        # Creating Production stage
        prod = HirastageStack(self, "prod")
        
        # Adding Test and commands
        unit_test=pipelines.ShellStep("unit_test", commands=["cd hira/","pip install -r requirements.txt",
                                                             "pip install -r requirements-dev.txt","pytest"])
        
        
        # Adding to the pipeline and  unit test as pre stage
        pipeline.add_stage(beta,pre=[unit_test])

        
        # Adding product to pipeline and manual approval as pre stage
        pipeline.add_stage(prod, pre = [pipelines.ManualApprovalStep("Approve prod stage")])


    # Assigning roles to Pipeline
    def create_pipeline_role(self):
            role = iam_.Role(self, "Hira-Pipeline-Role",
                   assumed_by=iam_.CompositePrincipal(
                       iam_.ServicePrincipal("lambda.amazonaws.com"),
                       iam_.ServicePrincipal("sns.amazonaws.com"),
                       iam_.ServicePrincipal("codebuild.amazonaws.com")),
                   managed_policies=[
                    iam_.ManagedPolicy.from_aws_managed_policy_name('CloudWatchFullAccess'),
                    iam_.ManagedPolicy.from_aws_managed_policy_name('AmazonDynamoDBFullAccess'),
                    iam_.ManagedPolicy.from_aws_managed_policy_name('service-role/AWSLambdaBasicExecutionRole'),
                    # iam_.ManagedPolicy.from_aws_managed_policy_name('AWSLambdaInvocation-DynamoDB'),
                    iam_.ManagedPolicy.from_aws_managed_policy_name("AwsCloudFormationFullAccess"),
                    iam_.ManagedPolicy.from_aws_managed_policy_name("AWSCodePipeline_FullAccess"),
                    iam_.ManagedPolicy.from_aws_managed_policy_name("AmazonS3FullAccess")
                    #iam_.ManagedPolicy.from_aws_managed_policy_name("SecretsManagerReadWrite")
                    ])
            return role

File: sprint5hira/lib/hirapipelinestack.ts
import { pipelines, Stack, StackProps} from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as cdk from 'aws-cdk-lib';
import { GitHubTrigger } from 'aws-cdk-lib/aws-codepipeline-actions';
import {Hirastage} from './hirastage'
import { ManualApprovalStep, ShellStep } from 'aws-cdk-lib/pipelines';
const app = new cdk.App();


export class Hirapipelinestack extends Stack {
    constructor(scope: Construct, id: string, props?: StackProps) {
      super(scope, id, props);

      
      // Using Secrets Manager to provide the access token to authenticate to GitHub
      const input = pipelines.CodePipelineSource.gitHub('hiraaziz2022skipq/sprint3Voyager', "main",{
                                            authentication:cdk.SecretValue.secretsManager('webtken'),
                                            trigger:GitHubTrigger.POLL,})

        /** 
        ShellStep()
                                            
        id -> 'synth'
        input -> source
        commands -> commands to run in pipeline
        primary_output_directory -> : Directory that will contain primary output fileset when script run
        
        **/

      const synth = new pipelines.ShellStep('Synth', {
                                            input: input,
                                            commands: ["cd sprint5hira","npm ci","npx cdk synth"],
                                            primaryOutputDirectory : "sprint5hira/cdk.out"})


      // Connecting to the github
      const pipeline = new pipelines.CodePipeline(this, 'Pipeline', {synth: synth});


        // Adding Test and commands
        const unit_test= new ShellStep("Unit_Test",{
        commands:["cd sprint5hira","npm ci","npm run test"]})


        // Instantiate Beta stage
        const stagebeta = new Hirastage(this,"betastage")
        // Adding beta stage to the pipeline and  unit test as pre stage
        pipeline.addStage(stagebeta,{pre:[unit_test]})


        // Creating Production stage
        const prod=new Hirastage(this,"prod")
        // Adding product to pipeline and manual approval as pre stage
        pipeline.addStage(prod,{pre:[new ManualApprovalStep("Waiting for your approval")]})}
  

  }